public/app/features/variables/editor/selectors.ts

```diff
@@ -0,0 +1,37 @@
+import {
+  AdHocVariableEditorState,
+  DataSourceVariableEditorState,
+  QueryVariableEditorState,
+  VariableEditorState,
+} from './reducer';
+
+export function getAdhocVariableState(
+  editorState: VariableEditorState
+): VariableEditorState<AdHocVariableEditorState> | null {
+  const { extended, ...rest } = editorState;
+
+  if (extended && 'dataSources' in extended) {
+    return {
+      extended,
+      ...rest,
+    };
+  }
+
+  return null;
+}
```

Comment:
Produces a _whole_ state object that matches a particular subtype. Ideally for use in a `mapStateToProps` function.

Reply:
(note I did [change this function](https://github.com/grafana/grafana/pull/44749/commits/eab9ecfb8ac6ddce46067227c4077dc9521db40a) slightly in a commit after this to return `VariableEditorState<null>` instead of just `null` when the type does not match)